arrow-libs/fx/arrow-collectors/src/commonMain/kotlin/arrow/collectors/Collectors.kt

```diff
@@ -0,0 +1,147 @@
+      override val characteristics: Set<Characteristics> = Characteristics.IDENTITY_CONCURRENT_UNORDERED
+      override suspend fun supply(): R = value
+      override suspend fun accumulate(current: R, value: Any?) {}
+      override suspend fun finish(current: R): R = current
+    }
+
+  /**
+   * Counts the number of elements in the flow.
+   */
+  public val length: Collector<Any?, Int> =
+    object : CollectorI<AtomicInt, Any?, Int> {
+      override val characteristics: Set<Characteristics> = Characteristics.CONCURRENT_UNORDERED
+      override suspend fun supply(): AtomicInt = AtomicInt(0)
+      override suspend fun accumulate(current: AtomicInt, value: Any?) {
+        current.incrementAndGet()
+      }
+
+      override suspend fun finish(current: AtomicInt): Int = current.get()
+    }
+
+  /**
+   * Sum of all the values in the flow.
+   */
+  public val sum: Collector<Int, Int> =
+    intReducer({ 0 }, Int::plus)
+
+  public fun intReducer(
+    initial: () -> Int, combine: (Int, Int) -> Int,
+  ): Collector<Int, Int> =
+    object : CollectorI<AtomicInt, Int, Int> {
+      override val characteristics: Set<Characteristics> = emptySet()
+      override suspend fun supply(): AtomicInt = AtomicInt(initial())
+      override suspend fun accumulate(current: AtomicInt, value: Int) {
+        current.update { combine(it, value) }
+      }
+
+      override suspend fun finish(current: AtomicInt): Int = current.get()
+    }
+
+  public fun <M> reducer(
+    initial: () -> M, combine: (M, M) -> M, unordered: Boolean = false,
+  ): Collector<M, M> =
+    object : CollectorI<Atomic<M>, M, M> {
+      override val characteristics: Set<Characteristics> =
+        if (unordered) Characteristics.CONCURRENT_UNORDERED
+        else emptySet()
+
+      override suspend fun supply(): Atomic<M> = Atomic(initial())
+      override suspend fun accumulate(current: Atomic<M>, value: M) {
+        current.update { combine(it, value) }
+      }
+
+      override suspend fun finish(current: Atomic<M>): M = current.get()
+    }
+
+  /**
+   * Returns the "best" value from the value.
+   *
+   * @param selectNew Decides whether the new value is "better" than the previous best.
+   */
+  public fun <M> bestBy(
+    selectNew: (old: M, new: M) -> Boolean,
+  ): Collector<M, M?> =
+    object : CollectorI<Atomic<M?>, M, M?> {
+      override val characteristics: Set<Characteristics> = Characteristics.CONCURRENT_UNORDERED
+      override suspend fun supply(): Atomic<M?> = Atomic(null)
+      override suspend fun accumulate(current: Atomic<M?>, value: M) {
+        current.update { old ->
+          if (old == null) value
+          else if (selectNew(old, value)) value
+          else old
+        }
+      }
+
+      override suspend fun finish(current: Atomic<M?>): M? = current.get()
+    }
+
+  /**
+   * Collects all the values in a list, in the order in which they are emitted.
+   */
+  public fun <T> list(): Collector<T, List<T>> =
```

Comment:
Can these become `val` properties, either single instance or computed?
Looks like the `supply` method provides a new instance each time, so I was wondering if the instance of `CollectorI` could just be cached or a singleton value.

Reply:
Unfortunately not, because `val`s cannot have generic arguments :(

Reply:
Idem as other comment, instance can be cached with `Any?` and casting. Same goes for everything else in this file.

Reply:
To add a bit more, if you try to make something like `list(): Collector<T, List<T>>` into a `val`, you run into the problem that `T` appears both in `in` and `out` position, so you cannot play the trick of using `Any?` or `Nothing` like in other places.

Reply:
Aha! Thanks for the trick, @nomisRev, this is fixed in a new commit